Adds Trinket Compare

Ellimistdev opened the pull request:
Needed a good way to see a trinket's impact across multiple specs.

### Adds: 
- Top level nav "Trinkets"
- New Trinket submenu to swap between item, item level, and fight styles

![trinkets_balefire_branch_454_castingpatchwerk](https://github.com/Bloodmallet/bloodmallet_web_frontend/assets/7306837/930900a5-f2d5-498a-90da-83a3f90b2b00)
![trinkets_caged_horror_486_castingpatchwerk3](https://github.com/Bloodmallet/bloodmallet_web_frontend/assets/7306837/5f0ef2a4-aeb6-4143-a349-b489b5a3cfb7)

### Outstanding todos:
MVP:
- Hook into db instead of fetching everything like a goblin
 - Need to add trinket-specific queries instead of needing to get ALL trinkets for ALL specs 😒 
- Server-side data handling

Midterm:
- Better trinket selection dropdown (scrolling at a minimum)
- Item_level handling for non-existent item levels for items
- Update fight_style in chart title.
- Trinket menu colors? Just picked Rogue.

Longterm:
- Swap item_name for item_id throughout to aid in localization
- Drilldown links: trinket_compare to trinkets

Bloodmallet commented:
This looks great! Let me add some information:
- the db does not "know" about simulation results and their inner workings. it is aware of where their jsons exist. therefore fetching all simulation results "like a goblin" is the way to go
- server side tries to be as uninvolved with game data as possible. it tries to let `bloodytools` in cooperation with `simc_support` do their thing and just pipe the result through to the frontend js

Midterm:

    Better trinket selection dropdown (scrolling at a minimum): [this](https://getbootstrap.com/docs/4.0/components/dropdowns/#options)  should hopefully help
    Item_level handling for non-existent item levels for items: when collecting trinket information each trinket should already know about which itemlevels are present for it. this can be used to filter the list or create it
    Update fight_style in chart title.: 👍 
    Trinket menu colors? Just picked Rogue.: I'd probably pick priest because it's the most neutral "color"

Longterm:

    Swap item_name for item_id throughout to aid in localization: 👍 
    Drilldown links: trinket_compare to trinkets: sounds great!

Bloodmallet commented on general_website/views/general.py:
I'd put all this logic into the chart importing javascript code instead of creating endpoints that basically self-crawl the website.

Ellimistdev commented:
Finally picked this back up! 

I reworked this so that it does not constantly self-crawl the site. Given that we currently receive all item data for any one spec, to generate these charts, we only need to hit each spec once and cache the result in localstorage (updated every 30 minutes). Additionally, I extracted several functions from `bloodmallet_chart_import()` to reduce complexity and improve reusability.

As far as I can tell, these changes don't impact existing functionality, though I have not ventured into accounts and custom charts.

![TrinketCompare](https://github.com/Bloodmallet/bloodmallet_web_frontend/assets/7306837/f45897c9-c110-4d79-8226-6a13dfd398e2)

Bloodmallet commented:
Sounds great! I'll need some time to review it. I'm currently busy preparing my own chart implementations and that's occupying my attention. I'll hopefully get back to you soon on this. 👍

Ellimistdev commented:
Updated for 11.1. 

Additional improvements: Added bm-utils for shared helpers, improved error handling, localization, caching, populating trinket compare menu with available items & item levels.

![TrinketCompare11 1](https://github.com/user-attachments/assets/a1900c31-058f-443b-909d-f66c22ecc768)

Ellimistdev commented:
@Bloodmallet ready for review :)

Ellimistdev commented on general_website/migrations/0008_alter_simulation_custom_apl_and_more.py:
Added automatically when following the development setup `python manage.py migrate`

Bloodmallet commented:
I'll try to check this out in more depth this coming weekend. They idea is still great!

Btw from a glance at your provided screenshots (thanks btw for those) was an itemlevel selection a specific consideration? Otherwise similar to the normal charts each trinket could straight up show all itemlevels. So the user would only need to select the trinket.

Ellimistdev commented:
> I'll try to check this out in more depth this coming weekend. They idea is still great!
> 
> Btw from a glance at your provided screenshots (thanks btw for those) was an itemlevel selection a specific consideration? Otherwise similar to the normal charts each trinket could straight up show all itemlevels. So the user would only need to select the trinket.

That's a great question. Initially, looking at a single item level across specs seemed like a smaller piece of work, which informed the data structure I'm using. 

![image](https://github.com/user-attachments/assets/97e5d40d-f687-475d-85a4-cfcedd0e4cf3)

Which is pretty different from the structure used by the class-specific trinket charts

![image](https://github.com/user-attachments/assets/600bf223-f4f0-4d01-914b-6a1aab678de5)

Ultimately, showing all iLvls simultaneously is a better user experience, but it wouldn't be a quick refactor.

Bloodmallet commented on general_website/static/general_website/js/bm-charts.js:
```diff
@@ -589,18 +818,11 @@ class BmChartData {
-            "pt_BR": "pt",
-            "ru_RU": "ru"
-        };
+        const subdomain = window.bmUtils.wowheadSubdomains;
```
I see why you did this and it absolutely makes sense to use another file to ensure not doubling the information. But this importer is [intended as a file other sites can import](https://bloodmallet.com/faq#data-usage) to show the charts on their own, relying on other files becomes a hassle. That's why I believe this can't get moved out.

Bloodmallet commented on general_website/static/general_website/js/bm-charts.js:
```diff
@@ -1611,79 +1839,101 @@ function bm_import_charts() {
+            console.debug(`Chart has chartId:`, chart_anchor.dataset.chartId);
             // if chart_id -> load id
-            let chart_id = chart_anchor.dataset["chartId"];
+            let chart_id = chart_anchor.dataset?.chartId;
```
Why would it be better here to basically create an invalid request with an `undefined` chart id instead of crashing the function?

Ellimistdev replied:
This goes back to safe async execution - If the DOM is still loading and the `chartId` is missing, we can retry `bm_import_charts()` without affecting the user experience. Without this, the application will hard crash and require a reload on the user's end (which might not resolve the missing attribute). This opens us up to gracefully handle missing charts if we want to e.g. on pages with multiple charts loading.

Bloodmallet commented on general_website/static/general_website/js/bm-charts.js:
```diff
@@ -1611,79 +1839,101 @@ function bm_import_charts() {
-            let wow_class = chart_anchor.dataset["wowClass"];
-            let wow_spec = chart_anchor.dataset["wowSpec"];
-            let chart_type = "trinkets";
+        } else if ("wowClass" in chart_anchor.dataset && "wowSpec" in chart_anchor.dataset &&
```
While I like this notation more I thought in js `in` wasn't as supported yet for this kind of checks?

Ellimistdev replied:
If you can believe it - `"key" in object` notation has been around since ES3 (the early 2000s)! Minor "gotcha" here is that `"prop" in obj` checks inherited props as well (which is especially good for browser code since we're just checking for the existence of the property without wanting to traverse the DOM) while `obj.hasOwnProperty("prop")` ignores inherited props.

Bloodmallet commented on general_website/static/general_website/js/bm-charts.js:
```diff
@@ -1611,79 +1839,101 @@ function bm_import_charts() {
+
             if (chart_anchor.dataset.hasOwnProperty("type")) {
-                chart_type = chart_anchor.dataset["type"];
+                chart_type = chart_anchor.dataset?.type;
```
I'm not familiar enough with js, so please bear with my questions, but what is this change improving?

Ellimistdev replied:
Good question! This pattern is called Optional Chaining. The original implementation `dataset["type"]` assumes that `chart_anchor.dataset` definitely exists, and will throw a runtime error if it doesn't. While `dataset?.type` will return undefined instead of crashing if `dataset` doesn't exist. This allows the code to safely execute if the data is available but the DOM isn't fully rendered - which is helpful in async execution. In fact, I should have removed the preceding line as well `if (chart_anchor.dataset.hasOwnProperty("type"))`, since that's implicit in optional chaining.
TL;DR: Cleaner and safer.

Bloodmallet commented on general_website/static/general_website/js/bm-charts.js:
```diff
@@ -1611,79 +1839,101 @@ function bm_import_charts() {
             request_endpoint = [endpoint, chart_type, fight_style, wow_class, wow_spec].join("/");
+        } else if ("type" in chart_anchor.dataset && chart_anchor.dataset.type === "trinket_compare") {
+            // Handle trinket_compare
+            item_name = chart_anchor.dataset?.itemName || "aberrant_spellforge";
```
Instead of hardcoding a fallback name here get the first trinket of the name-sorted list.

Bloodmallet commented on general_website/static/general_website/js/bm-charts.js:
```diff
@@ -1611,79 +1839,101 @@ function bm_import_charts() {
+        } else if ("type" in chart_anchor.dataset && chart_anchor.dataset.type === "trinket_compare") {
+            // Handle trinket_compare
+            item_name = chart_anchor.dataset?.itemName || "aberrant_spellforge";
+            item_level = chart_anchor.dataset?.itemLevel || "636";
```
Dynmically get the highest itemlevel as fallback.

Bloodmallet commented on general_website/static/general_website/js/bm-charts.js:
```diff
@@ -1611,79 +1839,101 @@ function bm_import_charts() {
+            // Handle trinket_compare
+            item_name = chart_anchor.dataset?.itemName || "aberrant_spellforge";
+            item_level = chart_anchor.dataset?.itemLevel || "636";
+            chart_type = chart_anchor.dataset?.type;
```
Haven't you already established in the `if` that this key exists?

Ellimistdev replied:
Great catch 🙏

Bloodmallet commented on general_website/static/general_website/js/bm-charts.js:
```diff
@@ -1611,79 +1839,101 @@ function bm_import_charts() {
+            if (chart_type === "trinket_compare") {
+                data = await getTrinketDataAsync(item_name, item_level, fight_style);
+            } else {
+                const response = await fetch(request_endpoint);
```
Why did you refactor the `XMLHttpRequest` to a `fetch` call?

Ellimistdev replied:
Many reasons! Essentially, `fetch` is "promise-based" which allows it to support `async/await` natively while `XMLHttpRequest` is "callback-based" and requires manual handling for async requests. [More info here](https://www.greatfrontend.com/questions/quiz/what-are-the-differences-between-xmlhttprequest-and-fetch)

Bloodmallet commented on general_website/static/general_website/js/bm-charts.js:
```diff
@@ -1611,79 +1839,101 @@ function bm_import_charts() {
-                    }
+            chart_anchor.dataset.loadedData = JSON.stringify(data);
+            let bm_data = new BmChartData(chart_anchor);
+            let chart = bm_data.data_type === "secondary_distributions" ? BmRadarChart : BmBarChart;
```
While I enjoy ternaries I wrote this withan if-conditional expecting more chart types to arrive in the future. While there are no immediate plans I thought it would be clearer which is loaded using such blocks instead of an ever growing long line of ternaries.

Bloodmallet commented on general_website/static/general_website/js/bm-charts.js:
```diff
@@ -3,6 +3,192 @@ const BmChartStyleUrl = "/static/general_website/css/bm-charts.css";
+const TRINKET_DATA_CACHE_KEY = 'trinketData';
+const TRINKET_DATA_CACHE_EXPIRY = 30 * 60 * 1000;  // 30 minutes in milliseconds
+
+const isTrinketDataCacheValid = () => {
```
What's the benefit of this function notation compared to the already present notation in the file?

Yours
```js
const functionName = (param1, param2) => {}
```

versus already present
```js
function functionName(param1, param2) {}
```

Ellimistdev replied:
Most modern frameworks prefer arrow functions over function declarations for cleaner and more consistent syntax, but there are also some major differences under the hood. Here, in particular, `function functionName(param)` does some stuff that we don't need, such as rebinding `this`, creating an `arguments` object, and hoisting the function.

Bloodmallet commented on general_website/static/general_website/js/bm-utils.js:
```diff
@@ -0,0 +1,220 @@
+ * @returns {string} The full language code (e.g., "en_US")
+ */
+const detectUserLanguage = (element = null) => {
+    let langCode = getLanguageFromCookie() || getLanguageFromDataset(element) || getLanguageFromBrowser();
```
If the website decided to set a language on the html element I believe this should have precedence over everything.

1. html element (dataset)
2. cookie
3. browser

Bloodmallet commented on general_website/static/general_website/js/bm-utils.js:
```diff
@@ -0,0 +1,220 @@
+    if (!text) return "Loading...";
+
+    switch (type) {
+        case "slug":
```
Is there a way to create an Enum with these options, that is also available when working on other files in VS Code?

Bloodmallet commented on general_website/static/general_website/js/navbar_trinket_menu.js:
```diff
@@ -0,0 +1,459 @@
+  const createInitialState = (overrides = {}) => ({
+    data_type: 'trinket_compare',
+    fight_style: 'castingpatchwerk',
+    wow_class: 'priest',
```
Why does this need a wow class at all?

Ellimistdev replied:
I suppose it doesn't 😁, this was to tap into the existing formatting elsewhere for coloring the text.

Bloodmallet commented on general_website/static/general_website/js/navbar_trinket_menu.js:
```diff
@@ -0,0 +1,459 @@
+   */
+  const fetchAvailableTrinkets = async (fightStyle) => {
+    try {
+      const fetchFunction = window.fetchAndProcessDataAsync ?? fetchAndProcessDataAsync;
```
What happens here?

Ellimistdev replied:
This line is using the "nullish coalescing operator" (`??`) to assign the first defined value between two options. If `window.fetchAndProcessDataAsync` is defined it gets used, otherwise it falls back to the local `fetchAndProcessDataAsync` function. 
This allows users to inject their own data handler if they're so inclined and opens up the ability for us to mock this code later on for testing. Probably overkill tbh

Bloodmallet commented on general_website/static/general_website/js/navbar_trinket_menu.js:
```diff
@@ -0,0 +1,459 @@
+  const fetchAvailableTrinkets = async (fightStyle) => {
+    try {
+      const fetchFunction = window.fetchAndProcessDataAsync ?? fetchAndProcessDataAsync;
+      const data = await fetchFunction?.(fightStyle);
```
It gets wilder and wilder for me. I thought `?.` allowed access to an object and in case this object doesn't have the follow up key it'll automatically return `undefined`. Now this calls...a function? Is this some sort of implicit error handling? So if the call fails/the function call throws an error, this instread will produce `undefined` again?

Ellimistdev replied:
Yes and no. There's no implicit error handling here from inside the function, but it does do type checking before calling the function. 
It is equivalent to 
```js
let data;
if (typeof fetchFunction === 'function') {
  data = await fetchFunction(fightStyle);
} else {
  data = undefined;
}
```
Errors thrown inside the function will be caught in the enclosing `try/catch`

Bloodmallet commented on general_website/static/general_website/js/navbar_trinket_menu.js:
```diff
@@ -0,0 +1,459 @@
+    try {
+      const fetchFunction = window.fetchAndProcessDataAsync ?? fetchAndProcessDataAsync;
+      const data = await fetchFunction?.(fightStyle);
+      return processTrinketsFromData(data);
```
In case the above assumption is correct I believe the code should already inform the user here about the missing data. Cause we're encountering the problem here. We expect data but don't get any. If we use `undefined` to create a chart it's suddenly up to the user to figure out where the weird chart creation call came from without data. At least to me that would be a suboptimal debugging experience.

Ellimistdev replied:
Absolutely right. While `processTrinketsFromData` validates the structure and contents of the response, logs errors, and returns a safe fallback (`[]`), there is still a gap where we need to validate that `data` is non-empty before passing it forward.

Bloodmallet commented on general_website/static/general_website/js/navbar_trinket_menu.js:
```diff
@@ -0,0 +1,459 @@
+      const userLanguage = window.bmUtils?.detectUserLanguage() || 'en_US';
+  
+      for (const trinketKey in data.items) {
+        if (trinketKey === "baseline") continue;
```
I know this works. I know some people love to not use parenthesis for such cases. I don't. I like to be explicit about what is the intended part of the conditional. And In case this section grows the diff will only add elements within the already prepared block and not "alter" the if-condition line.

```suggestion
        if (trinketKey === "baseline") {
          continue;
        }
```

Bloodmallet commented on general_website/static/general_website/js/navbar_trinket_menu.js:
```diff
@@ -0,0 +1,459 @@
+        if (trinketKey === "baseline") continue;
+  
+        let trinketName = trinketKey.replace(/_/g, ' ')
+          .replace(/\b\w/g, c => c.toUpperCase()); // Default formatting
```
Wait...didn't you add a formating fucntion for such a case?

Bloodmallet commented on general_website/static/general_website/js/navbar_trinket_menu.js:
```diff
@@ -0,0 +1,459 @@
+        });
+      }
+  
+      return availableTrinkets.sort((a, b) => a.name.localeCompare(b.name));
```
Oh nice. I didn't know about localCompare.

Bloodmallet commented on general_website/static/general_website/js/navbar_trinket_menu.js:
```diff
@@ -0,0 +1,459 @@
+      console.debug(`Loaded ${availableTrinkets.length} available trinkets`);
+  
+      // Set up observer to watch for future changes
+      const observer = new MutationObserver(async (mutations) => {
```
I need to read up on `MutationObserver`

Ellimistdev replied:
`MutationObserver` is so cool 😁This avoids manually calling a chart update function after each fetch and just reacts to when the data in the chart actually changes instead of looking for events or something

Bloodmallet reviewed (commented):
This. Is. Awesome!
Thank you very much for this PR. This looks like a lot of work and care was put into it. I added a few comments to better understand your intentions and what's going on. Please have a look.

I love there are several opportunities for me to learn a lot.

I'm very much looking forward to get this merged soon.

Bloodmallet commented:
I want to add that after reading your PR I want to look into applying the following constructs I've seen in your PR to the whole code base:

- localStorage
- MutationObserver
- fetch

```js
let a = get_something || default;
```

```js
let a = object?.magicalKey;
```

Registering functions (and classes?) to the window
```js
window.customkey = {
    function1,
    function2,
    class1
};
```

Your `bmUtils.createElement` versus creating all elements "by hand" as is currently done everywhere else.

Thank you very much

Ellimistdev commented:
Cleaning up [my branch ](https://github.com/Ellimistdev/bloodmallet_web_frontend/tree/bmRefactor) and will push changes this week, but got the multilevel trinket display working:
![multiilvl](https://github.com/user-attachments/assets/dc774bc8-0ffa-4008-96e7-84bd1e763bec)

Also did a decent amount of refactoring w.r.t. chart/component creation 
 1. Modular Architecture

Before: Everything was in a single monolithic file with global functions
After: Clear separation into classes (BmChartData, BmBarChart, BmRadarChart, BmUIUtils, BmChartComponents)
Value: Better maintainability, testability, and code organization

2. Utility Class Introduction

New: BmUIUtils class with comprehensive helper methods
Methods Added:
- Language detection (detectUserLanguage, getLanguageFromCookie, etc.)
- Text formatting (formatText, formatNumber, capitalizeFirstLetters)
- DOM creation (createElement, createDiv, createSpan)
- Tooltip system integration
- Safe JSON parsing

Value: Centralized utilities, reduced code duplication, better error handling

3. Component-Based Chart Creation

New: BmChartComponents class for chart element creation
Methods: createLegend, createAxisTitles, createChartRow, createTooltip, etc.
Value: Modular chart building, easier customization and testing

Code Quality Improvements
4. Better Documentation

Before: Minimal JSDoc comments
After: Comprehensive JSDoc with parameter types, return types, and examples
Value: Better developer experience, easier onboarding

5. Static Methods and Organization

Before: Global functions scattered throughout
After: Logical grouping in static methods within classes
Value: Better namespace management, clearer API

6. Enhanced Filter System

New: Static methods for filtering (passesActivePassiveFilter, passesItemLevelFilter, compareDPSValues)
Value: More maintainable filtering logic, easier to extend

Bloodmallet commented:
Thank you for the updates. I'm scheduling this for review next weekend. Hopefully I'll be able to also merge it by then. Looking forward to it!

Bloodmallet commented:
This looks incredible even without your announced other changes! So many things to learn for me in it. Thank you very much.

I'm going ahead and merge this already. You worked hard on this for such a long time. It's time to release it! Please open a new PR for the additional changes you announced above.
